### DISCUSSION TITLE
How to properly clean up multiple levels of spawned actors?

### DISCUSSION THREAD

**Discussion by ezhikov** [statelyai/xstate#1558] [Q&A] [answered]
Hi. I have multiple machines that spawn each other, and I'm not sure how to properly clean it up on all levels. If I understand correctly, difference between invocation and spawning is that with invocation we're getting automatic cleanup when we exit state, and we have to clean up actors by hand. Next difference is that we have to know amount of _invoked_ actors beforehand and we can _spawn_ arbitrary amount of actors anytime.

What bothers me here, is how to properly cleanup all those spawned actors. I have some top level machine. This machine can spawn multiple actors, that, in order will spawn 3 to 15 actors each. I wonder, if I should stop every actor from the very bottom (with passing a message, perhaps?), cleanup context, then go level up and perform same manipulations on another set of actors?

And, as a side question which could lead me to an easier solution: is it possible to invoke arbitrary amount of child machines? Something like `invoke: (ctx, event) => [{src: 'child', src: 'child', src: 'child'}]`

**Comment by Andarist** [comment, marked as answer]
> I wonder, if I should stop every actor from the very bottom (with passing a message, perhaps?), cleanup context, then go level up and perform same manipulations on another set of actors?

If you want to stop an entire "branch" of actors you only need to stop the one at the top and it will propagate that recursivelyyto all of its descendants.

> And, as a side question which could lead me to an easier solution: is it possible to invoke arbitrary amount of child machines? Something like invoke: (ctx, event) => [{src: 'child', src: 'child', src: 'child'}]

It should be possible, although a different syntax might be needed - `invoke` should accept an array of invoke configs.

**Reply by ezhikov** [reply]
> you only need to stop the one at the top
That sounds neat, if I understand correctly. 

So, if this is my setup, when my `parent` goes back to `off` state it will cleanup child, since it invoked, and grandchildren and descendants will automatically be cleaned up?
```js
const grandchild = Machine({
  id: "grandchild",
  intial: "idle",
  entry: ["spawn_more_descendants"],
  states: {
    idle: {},
  },
});

const child = Machine({
  id: "child",
  initial: "idle",
  entry: ["spawn_grand_children"],
  states: {
    idle: {},
  },
});


const parent = Machine({
  id: "parent",
  initial: "turned_off",
  states: {
    turned_off: {
      on: { turn_on: "turned_on" },
    },
    turned_on: {
      on: {
        turn_off: "turned_off",
      },
      invoke: {
        src: () => child,
      },
    },
  },
});
```

> invoke should accept an array of invoke configs.

But I can't determine size of this array and machine ids in runtime based on context or event. I have to specify it beforehand.
